revision of color line section (#135)
googlefonts/colr-gradients-spec: colr-gradients-spec.md

> 

Reviewer: Generally, please try to line-break the markdown to 80 columns if possible.

> 

Reviewer: Generally, please try to line-break the markdown to 80 columns if possible.

**5.7.11.1.2.1 Color Lines**

A color line is a function that maps real numbers to color values to define a one-dimensional gradation of colors, to be used in the definition of linear or radial gradients. A color line is defined as a set of one or more color stops, each of which maps a particular real number to a specific color.

> A color line is not a line in a geometric sense: on its own, a color line has no positioning, orientation or size within a design grid. These geometric aspects are part of the definition of a linear or radial gradient. Specifically, the gradient definition will specify positions in the design grid that correspond to the real values 0 and 1 in the color line, with placement in the design grid for other numeric values of the color line interpolated from the placement of 0 and 1.

Reviewer: As is, contrasting it to a geometric line, I find this to be rather having the character of a note. Perhaps we can rephrase positively:

```suggestion
On its own, a color line has no positioning, orientation or size within a design grid. By mapping and referencing a color line, linear and radial gradient definitions will specify positions in the design grid that correspond to the real values 0 and 1 in the color line, as well as rules for where to draw interpolated colors of the color line interpolated that follow from the placement of color stops 0 and 1.
```

> A color stop is defined by a real number, the *stop offset*, and a color. A color line is defined with at least one color stop within the interval [0, 1]. Additional color stops can be specified within or outside the interval [0, 1]. (Stop offsets are represented using F2DOT14 values, therefore color stops can only be specified within the range [-2, 2). See <span style="color:red">5.7.11.2.x</span> for format details.) If only one color stop is specified, that color is used for the entire color line; at least two color stops are needed to create color gradation.

Reviewer: > A color line is defined with at least one color stop within the interval [0, 1].

Filed #148 for that, I don't think we need to enforce that, but we can handle that separately.

> Suppose a color line has two or more color stops, and consider these to be ordered in increasing stop offset value. Color gradation is defined over the interval from the first color stop, through the successive color stops, to the last color stop. Between consecutive color stops, color values are interpolated between the colors of the two stops.

Reviewer: Can we generalize this and avoid the specific "Suppose...", to e.g.:
Between adjacent color stops, at any position between the two stops, the color is the linearly interpolated color value at that position.

> > **_TBD: Does interpolation of colors need further specification?_**

Reviewer: Minor: I prefer to file issues instead of inline TODOs.

Author: Agree we should have issues, but I think it's helpful to call out in the draft text so that the open issues are less likely to be overlooked.

Author: Opened #150

Reviewer: Thanks for opening the issue - one reason I find inline TBDs somewhat redundant is because we may also find in the discussion of the issue that we do not need further clarifications. Then we can just close the issue. But for removing the inline TBDs we'd need another edit. My preference would still be to remove them inline. @rsheeter, @anthrotype - opinions?

If there are multiple color stops defined for the same stop offset, the first one is used for computing color values on the color line below that stop offset, and the last one is used for computing color values at or above that stop offset. All other color stops for that stop offset are ignored.

While the color gradation is specified over a defined interval, the color line continues indefinitely outside that interval in both directions. The color pattern outside the defined interval is repeated according to the color line’s *extend mode*. Three extend modes are supported:

> * Pad: outside the defined interval, the color of the closest color stop is used. Using a string as an analogy, given a sequence “ABC”, it is extended to “…*AA* ABC *CC*…”.

Reviewer: Formatting suggestion, and an optional suggestion to use less programming-specific terminology:

```suggestion
* **Pad:** outside the defined interval, the color of the closest color stop is used. Using a sequence of letters as an analogy, given a sequence “ABC”, it is extended to “…*AA* ABC *CC*…”.
```

  defined in CSS
  [linear-gradient()](https://developer.mozilla.org/en-US/docs/Web/CSS/linear-gradient)
  and
  [radial-gradient()](https://developer.mozilla.org/en-US/docs/Web/CSS/radial-gradient)
  functions.
While the color gradation is specified over a defined interval, the color line continues indefinitely outside that interval in both directions. The color pattern outside the defined interval is repeated according to the color line’s *extend mode*. Three extend modes are supported:

* a repeated gradient along the gradient positions (linear or radial): divide 1
  by the number of desired repetitions and use the result as your maximum color
  stop, then use color line extend mode *repeat* to have it continue outside the
  defined interval.
* Pad: outside the defined interval, the color of the closest color stop is used. Using a string as an analogy, given a sequence “ABC”, it is extended to “…*AA* ABC *CC*…”.

* a mirrored / color-circle gradient: divide 1 by two times the number of
  desired full color stripes, and define the color stops between the 0 and the
  result of this division, then use color line extend mode *reflect* to have it

>   continue mirrored.

Reviewer: This type of cookbook section is not contained in the new text and I find it helpful to keep as it illustrates the utility of defining ranges for sub ranges of [0,1]. Can we add it back as a suitable subsection?

Author: The wording has a lot of assumptions and wasn't clear. But I'll work on clarifying what I think was the intent.

Reviewer: Thank you for adding this section back in. I find it not helpful to read blanket judgements applied like "wasn't clear" / "had a lot of assumptions" without stating what exactly was not clear or which assumptions should have been called out explicitly. Similarly, in my opinion a commit description stating "with a clearer, more accurate and more complete description." is broad statement potentially not shared by all readers. Our perceptions about the clarity, precision or meaning of a text can individually differ. 

In this case of the cookbook section, I personally found the the previous text just as clear as the new one. I would prefer that we talk about what specifically did not seem clear and then discuss, iterate the wording and improve it so we arrive at spec text that is most broadly understood.

> * Repeat: The color line is repeated over repeated multiples of the defined interval. For example, if color stops are specified for defined interval of [0, 0.7], then the pattern is repeated above the defined interval for intervals [0.7, 1.4], [1.4, 2.1], etc.; and also repeated below the defined interval for intervalues [-0.7, 0], [-1.4, -0.7], etc. In each repeated invterval, the first color is that of the farthest defined color stop. By analogy, given a sequence “ABC”, it is extended to “…*ABC* ABC *ABC*…”.

Reviewer: Formatting, typos.

```suggestion
* **Repeat:** The color line is repeated over repeated multiples of the defined interval. For example, if color stops are specified for defined interval of [0, 0.7], then the pattern is repeated above the defined interval for intervals [0.7, 1.4], [1.4, 2.1], etc.; and also repeated below the defined interval for intervals [-0.7, 0], [-1.4, -0.7], etc. In each repeated interval, the first color is that of the farthest defined color stop. By analogy, given a sequence “ABC”, it is extended to “…*ABC* ABC *ABC*…”.
```

> * Reflect: The color line is repeated over repeated intervals, as for the repeat mode. However, in each repeated interval, the ordering of color stops is the reverse of the adjacent interval. By analogy, given a sequence “ABC”, it is extended to “…*ABC CBA* ABC *CBA ABC*…”.

Reviewer: ```suggestion
* **Reflect:** The color line is repeated over repeated intervals, as for the repeat mode. However, in each repeated interval, the ordering of color stops is the reverse of the adjacent interval. By analogy, given a sequence “ABC”, it is extended to “…*ABC CBA* ABC *CBA ABC*…”.
```

**5.7.11.1.2.1 Color Lines**

A color line is a function that maps real numbers to color values to define a
one-dimensional gradation of colors, to be used in the definition of linear or
radial gradients. A color line is defined as a set of one or more color stops,
each of which maps a particular real number to a specific color.

> On its own, a color has no positioning, orientation or size within a design

Reviewer: ```suggestion
On its own, a color line has no positioning, orientation or size within a design
```

I think `line` i missing here, otherwise the paragraph reads well now, thanks.
